### PR TITLE
Javascript2 week2

### PR DESCRIPTION
**Pull request by olhasashchuk**
- Previous comments from mentor implemented
- It was implemented table with data of currency, where data change dynamically when a new rate is added.
- it was added a simple search function.

### CODE REVIEW

**Comment by olhasashchuk** [comment]
Deploy site: https://currencyconvert-sava.netlify.app/

File: index.html [hunk 1]
```diff
@@ -11,47 +11,107 @@
...
+
+      <div class="section_create_rate">
          <div>
-            <h2 class="subtitle">Insert a new currency rate and add a new rate</h2>
-            <form class="form p-4 formRate--js">
+            <h2 class="subtitle">Insert a new currency rate</h2>
+            <form class="form p-4 formRate--js">            
                <div class="form-group pb-3">
                   <label for="currency_from">From:</label>
-                  <input class="form-control" id="currency_from" type="text" name="currency_from" placeholder="EUR" required>
```
Comment (islanderupnorth) [inline index.html:19]: It's great you have used a CSS framework like Bootstrap to style your HTML code.

Make sure to be consistent in your naming conventions. Standard practice for HTML is `kebab-case` for HTML/CSS and not `snake_case`. It looks like you have a combination for `camelCase` i.e. `formRate--js`, `kebab-case` & `snake_case`.

File: script.js [hunk 2]
```diff
@@ -1,60 +1,274 @@
...
+   createTableHeaders(tableCurrencyCol, addedColumns);
+   createTableRows(tableCurrencyRow, addedColumns);
+}
+
+createTable();
+
+// Function to populate select elements for currency rates
+function populateSelectElements(currencyElements, addedCurrencies, isBase = false) {
+   arrRates.forEach(item => {
+      const currencyKey = isBase ? [item.base] : Object.keys(item.rates);
```
Comment (islanderupnorth) [inline script.js:74]: Nice use case of the ternary operator

File: script.js [hunk 3]
```diff
@@ -1,60 +1,274 @@
...
+   document.querySelector('.formConvert--js').addEventListener('input', function(event) {
+      event.preventDefault();
+      getConvert();
+   });
+
+   document.querySelector('.formSearchRate--js').addEventListener('submit', function(event) {
+      event.preventDefault();
+      searchRate();
+   });
+}
```
Comment (islanderupnorth) [inline script.js:125]: Really great how you have compartmentalised your code in general into smaller functions

File: script.js [hunk 4]
```diff
@@ -1,60 +1,274 @@
...
+function createRates() {
+   const selectCurrencyFrom = document.querySelectorAll('.select_currency_from');
+   const selectCurrencyTo = document.querySelectorAll('.select_currency_to');
+   if (!selectCurrencyFrom.length || !selectCurrencyTo.length) return;
+
+   const addedCurrenciesFrom = new Set();
+   const addedCurrenciesTo = new Set();
+
+   populateSelectElements(selectCurrencyFrom, addedCurrenciesFrom, true);
+   populateSelectElements(selectCurrencyTo, addedCurrenciesTo, false);
```
Comment (islanderupnorth) [inline script.js:97]: Nice and re-useable

File: script.js [hunk 5]
```diff
@@ -1,60 +1,274 @@
...
+   const newRate = parseFloat(document.getElementById('update-currency_rate').value);
+   const updateCurrencyFrom = document.getElementById('update-currency_from').value;
+   const updateCurrencyTo = document.getElementById('update-currency_to').value;
+
+   let existingRate = arrRates.find(rateObj => rateObj.base === updateCurrencyFrom);
+   if (existingRate) {
+      existingRate.rates[updateCurrencyTo] = newRate;
+      addDataColTableCurrency(updateCurrencyFrom, updateCurrencyTo, newRate);
+   }
+}
```
Comment (islanderupnorth) [inline script.js:240]: From a UX perspective then I would actually call this function when changes are made to the "From" and "To" currency selectors. As soon as I change currency it automatically updates the new currency conversion.

Also on the "amount" input field if string is empty don't updateRate() which results in a `NaN` result

Minor UX improvements

File: index.html [hunk 6]
```diff
@@ -11,47 +11,107 @@
...
+                  <input class="form-control currencyName_from" id="currency_from" type="text" name="currency_from" placeholder="EUR" required>
                </div>
                <div class="form-group pb-3">
                   <label for="currency_to">To:</label>
-                  <input class="form-control" id="currency_to" type="text" name="currency_to" placeholder="USD" required>
+                  <input class="form-control currencyName_to" id="currency_to" type="text" name="currency_to" placeholder="USD" required>
```
Comment (islanderupnorth) [inline index.html:88]: Both `currencyName_from` and `currencyName_to` update naming convention and add `text-transform: uppercase;` to these elements so they are by default capitalised.

**Review by islanderupnorth — COMMENTED** [review]
Great job Olha - solid Javascript fundamentals!

Really great with the added bonus of using an external library to help style your form! I personally am a big fan of something called [Tailwindcss](https://tailwindcss.com/) which is essentially the latest _"hot"_ framework which is very similar to Bootstrap which is a solid framework that has been around for a while.

In general my main feedback which is easy to fix is be wary of being consistent in your naming conventions. Other than that you are really good at abstracting logic into smaller functions so keep up that habbit!

**Comment by olhasashchuk** [comment]
> Great job Olha - solid Javascript fundamentals!
> 
> Really great with the added bonus of using an external library to help style your form! I personally am a big fan of something called [Tailwindcss](https://tailwindcss.com/) which is essentially the latest _"hot"_ framework which is very similar to Bootstrap which is a solid framework that has been around for a while.
> 
> In general my main feedback which is easy to fix is be wary of being consistent in your naming conventions. Other than that you are really good at abstracting logic into smaller functions so keep up that habbit!

Thank you so much your response! I will look at all of your comments and correct them. I used Tailwaild in the my previous project. And I will improve this project next time. Thanks for your recommendation!

**Pull request merged**